### PR TITLE
[ENH] Improve deep learning networks test coverage for parameters which can be list

### PR DESCRIPTION
**Pull request by Cyril-Meyer**
#### Reference Issues/PRs

Improves part of #1233

#### What does this implement/fix? Explain your changes.

Improve the `test_all_networks_functionality` tests, adding cases when some parameters can be a list.

| Before | After |
|:-:|:-:|
|before|after2|

#### Does your contribution introduce a new dependency? If yes, which one?

No.

#### Any other comments?

Some tests won't work and are disabled as 'continue exceptions'. I don't know if this is the right way to do it.
In any case, these problems will need to be fixed, and are related to the implementations themselves.
Anyway, we are adding new tests, and these cases are just like before, without the improvement.

```python
# Exceptions to fix
if (
    attrname in ["kernel_size", "padding"]
    and network.__name__ == "TapNetNetwork"
):
    continue
if attrname in ["activation"] and network.__name__ == "AEBiGRUNetwork":
    continue
# LITENetwork does not seem to work with list args
if network.__name__ == "LITENetwork":
    continue
```

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

### CODE REVIEW

File: aeon/networks/tests/test_all_networks.py [hunk 1]
```diff
@@ -41,20 +41,65 @@ def test_all_networks_functionality(network):
...
+                "dilation_rate",
+                "use_bias",
+            ]:
+                # Exceptions to fix
+                if (
+                    attrname in ["kernel_size", "padding"]
+                    and network.__name__ == "TapNetNetwork"
+                ):
+                    continue
+                if attrname in ["activation"] and network.__name__ == "AEBiGRUNetwork":
```
Comment (hadifawaz1999) [inline aeon/networks/tests/test_all_networks.py:61]: can skip the "AE" networks as they will be having their own tests for special reasons (them being auto encoders)

**Comment by hadifawaz1999** [comment]
Thanks for taking care of this, coverage seems great now !

We should open an issue after this getting merged for cases like LITE (still not parametrized as it should with list inputs)

File: aeon/networks/tests/test_all_networks.py (hunk 1 above)
Comment (Cyril-Meyer) [reply aeon/networks/tests/test_all_networks.py]: Skip added in b99b90df39eff26877ee7cb808ecca5ceda8b073 with deletion of all other "AE" models checks.

Comment (MatthewMiddlehurst) [reply aeon/networks/tests/test_all_networks.py]: This should use the tag instead of the network name IMO

**Review by MatthewMiddlehurst — COMMENTED** [review]
This should be a separate test IMO i.e. `test_all_networks_params`.

Will these parameters be in every network (even future ones)? May be missing something, but would this fail if a network does not have one from the list?

File: aeon/networks/tests/test_all_networks.py (hunk 1 above)
Comment (hadifawaz1999) [reply aeon/networks/tests/test_all_networks.py]: i agree @MatthewMiddlehurst but issue is that we still havent added the config to all networks, only the ones added with clustering project for AE based

**Comment by hadifawaz1999** [comment]
> This should be a separate test IMO i.e. `test_all_networks_functionality`.
> 
> Will these parameters be in every network (even future ones)? May be missing something, but would this fail if a network does not have one from the list?

hmm good point @MatthewMiddlehurst , maybe we can have a test per network ? test_inception_parameters, test_lite_parameters etc.

File: aeon/networks/tests/test_all_networks.py (hunk 1 above)
Comment (MatthewMiddlehurst) [reply aeon/networks/tests/test_all_networks.py]: seems like an easy change given they all start with AE anyway?

Comment (Cyril-Meyer) [reply aeon/networks/tests/test_all_networks.py]: @MatthewMiddlehurst we can use tags, but if we want to, we have to add a check that the networks has tags before.
```python
if network.__name__.startswith("AE")
```
can be replaced by something like
```python
if hasattr(network, "_config") and network._config.structure == "auto-encoder"
```

**Comment by Cyril-Meyer** [comment]
> This should be a separate test IMO i.e. `test_all_networks_params`.

Not sure about this as it's exactly the same test as before but with different types for some parameters.
But yeah, it's easy to change as you asked :)

> hmm good point @MatthewMiddlehurst , maybe we can have a test per network ? test_inception_parameters, test_lite_parameters etc.

This is also a valid choice, but which requires a new test to be created for each new network.
For now, I was mimicking the actual tests that do not require that.

> Will these parameters be in every network (even future ones)? May be missing something, but would this fail if a network does not have one from the list?

The networks do not share the parameters, that's why we check that the network have the parameters or not. So no, this would not fail if a network does not have the parameter.
The only reason to have a fail is if a new network use the same parameters name but do not allow the usage of both types (e.g. like LITE for now).

**Comment by MatthewMiddlehurst** [comment]
I do not mind it being a singular test for now instead of going through each file and adding a bunch of tests. Ideally the parameters would be used in more than one network though.

> Not sure about this as it's exactly the same test as before but with different types for some parameters.
> But yeah, it's easy to change as you asked :)

I think it's acceptable to lose a bit of testing efficiency here for readability. You don't have to assert the output etc like the current test. I would like to avoid forming the "omni-test" for all networks which just covers everything in a single function, it was kinda like that before 🙂.

**Comment by hadifawaz1999** [comment]
> I added the new test as a new function, `test_all_networks_params` as proposed. I also changed the AE skip using tag instead of the network name and added a skip info.
> 
> Also, is `tensorflow-addons` requirement still needed ? [ref](https://github.com/aeon-toolkit/aeon/blob/c980d7f7cb631672f16396f37691825e0743c4ad/aeon/networks/_encoder.py#L49)

thanks for the changes @Cyril-Meyer ! the check for config existence is good solution for now, we should keep in mind removing it once we add the config to all other networks

great catch on tensorflow-addons, we dont need it no, we removed its usage before and project  no longer depends on it, probably forgot to remove it here

**Comment by Cyril-Meyer** [comment]
> thanks for the changes @Cyril-Meyer ! the check for config existence is good solution for now, we should keep in mind removing it once we add the config to all other networks

The check was not necessary, as the base class has it :
```python
    _config = {
        "python_dependencies": ["tensorflow"],
        "python_version": "<3.12",
        "structure": "encoder",
    }
```
I just didn't see this before, but it's cool.


> great catch on tensorflow-addons, we dont need it no, we removed its usage before and project no longer depends on it, probably forgot to remove it here

Yeah, I think it will be more clear if we create a new issue and solve it with a new PR (I can handle that).

**Comment by hadifawaz1999** [comment]
Looks good to me @MatthewMiddlehurst how about you ? (btw for some reason readthedocs says "pending" for 12 hours now)

**Review by MatthewMiddlehurst — APPROVED** [review]
Looks fine for now. Feel free to merge when you think its good.

**Comment by MatthewMiddlehurst** [comment]
Can ignore the docs, says it has built fine when you click on details.

**Review by hadifawaz1999 — APPROVED** [review]
LGTM! thanks for taking care of this

**Pull request merged**